source/Coop/Mod/Persistence/RPC/ArgumentFactory.cs

```diff
@@ -1,60 +1,121 @@
+                        arg.StoreObjectId.Value,
+                        resolvedObject,
+                        resolvedObject.GetType());
+                    store.Remove(arg.StoreObjectId.Value);
+                    return resolvedObject;
                 default:
                     throw new ArgumentOutOfRangeException();
             }
         }
 
-        public static object[] Resolve(this RailClientRoom room, List<Argument> args)
+        /// <summary>
+        ///     Resolves a list of transferred RPC arguments to be used in the local function call.
+        /// </summary>
+        /// <param name="room">Rail room for the local client.</param>
+        /// <param name="store">Clients remote store instance.</param>
+        /// <param name="args">Argument to be resolved.</param>
+        /// <returns>A list of the unwrapped arguments.</returns>
+        public static object[] Resolve(IStore store, List<Argument> args)
         {
-            return args.Select(arg => room.Resolve(arg)).ToArray();
+            return args.Select(arg => Resolve(store, arg)).ToArray();
         }
 
-        public static Argument Create(object obj)
+        /// <summary>
+        ///     Creates a RPC transfer wrapper for a function call argument.
+        /// </summary>
+        /// <param name="store">Clients remote store instance.</param>
+        /// <param name="obj">The object to be wrapped.</param>
+        /// <param name="bTransferByValue">
+        ///     If true, the object will always be transferred by
+        ///     value. If the argument is too large to fit in an <see cref="RailEvent" />,
+        ///     the argument will be shared to all receivers using the <paramref name="store" />.
+        ///     If false, the argument may be, depending on the type, transferred by reference.
+        /// </param>
+        /// <returns>The wrapped argument.</returns>
+        public static Argument Create(
+            [NotNull] IStore store,
+            [CanBeNull] object obj,
+            bool bTransferByValue)
         {
             switch (obj)
             {
                 case null:
                     return Argument.Null;
+                case MBObjectManager mbObjectManager:
+                    return Argument.MBObjectManager;
                 case MBGUID guid:
                     return new Argument(guid);
-                case RailEntityBase entity:
-                    return new Argument(entity);
-                case EntityId entityId:
-                    return new Argument(entityId);
-                case MBObjectBase mbobj:
-                    return new Argument(mbobj.Id);
                 case int i:
                     return new Argument(i);
+                case float f:
+                    return new Argument(f);
+                case MBObjectBase mbobj:
+                    if (bTransferByValue)
+                    {
+                        goto default;
```

Comment:
I hesitate seeing goto statements. I recommend having the return in the if statement and having the code execute past the case on false instead.